File: scripts/extractAllCivilCodeFromPDF.mjs
/**
 * Script d'extraction de tous les articles du Code Civil Algérien depuis le PDF
 * Extrait le texte exact tel qu'il apparaît dans le PDF
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import pdfParse from 'pdf-parse';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Chemins
const PDF_PATH = path.join(__dirname, '..', 'data', 'Code_Civil.pdf');
const OUTPUT_PATH = path.join(__dirname, '..', 'data', 'civilCodeComplete.ts');

// Mots-clés pour identifier les articles pertinents pour la sécurité sociale
const SS_KEYWORDS = [
  'responsabilité', 'dommage', 'préjudice', 'indemnisation', 'réparation',
  'incapacité', 'invalidité', 'accident', 'travail', 'maladie',
  'professionnel', 'employeur', 'salarié', 'victime', 'consolidation',
  'expertise', 'médical', 'frais', 'soins', 'traitement', 'tierce personne',
  'esthétique', 'moral', 'corporel', 'faute', 'négligence', 'imprudence',
  'assurance', 'cotisation', 'prestation', 'pension', 'rente'
];

/**
 * Extrait le texte du PDF
 */
async function extractTextFromPDF() {
  console.log('📖 Lecture du PDF:', PDF_PATH);
  
  const dataBuffer = fs.readFileSync(PDF_PATH);
  const data = await pdfParse(dataBuffer);
  
  console.log(`✅ PDF lu: ${data.numpages} pages, ${data.text.length} caractères`);
  return data.text;
}

/**
 * Extrait les articles du texte
 */
function extractArticles(text) {
  console.log('\n🔍 Extraction des articles...');
  
  const articles = [];
  
  // Pattern pour détecter les articles
  // Format: "Article XXX" ou "Art. XXX" suivi du texte jusqu'au prochain article
  const articleRegex = /(?:Article|Art\.?)\s+(\d+(?:\s*bis|\s*ter)?)\s*[:\.\-]?\s*([\s\S]*?)(?=(?:Article|Art\.?)\s+\d+|$)/gi;
  
  let match;
  while ((match = articleRegex.exec(text)) !== null) {
    const number = match[1].trim();
    let content = match[2].trim();
    
    // Nettoyer le contenu
    content = content.replace(/\s+/g, ' ');  // Espaces multiples
    content = content.substring(0, 3000);  // Limiter à 3000 caractères
    
    if (content.length > 10) {  // Ignorer les articles vides
      const { isRelevant, category, keywords } = categorizeArticle(number, content);
      
      articles.push({
        number,
        content,
        category,
        keywords,
        relevantForSS: isRelevant
      });
    }
  }
  
  console.log(`✅ ${articles.length} articles extraits`);
  return articles;
}

/**
 * Catégorise un article et détermine sa pertinence pour la SS
 */
function categorizeArticle(number, content) {
  const contentLower = content.toLowerCase();
  
  // Compter les mots-clés présents
  const presentKeywords = SS_KEYWORDS.filter(kw => contentLower.includes(kw));
  const relevanceScore = presentKeywords.length;
  
  // Seuil de pertinence: au moins 2 mots-clés
  if (relevanceScore >= 2) {
    let category = 'Général';
    
    if (['responsabilité', 'faute', 'négligence', 'imprudence'].some(w => contentLower.includes(w))) {
      category = 'Responsabilité';
    } else if (['indemnisation', 'réparation', 'dommage', 'préjudice'].some(w => contentLower.includes(w))) {
      category = 'Indemnisation';
    } else if (['expertise', 'consolidation', 'procédure'].some(w => contentLower.includes(w))) {
      category = 'Procédure';
    } else if (['assurance', 'cotisation', 'prestation'].some(w => contentLower.includes(w))) {
      category = 'Sécurité Sociale';
    }
    
    return {
      isRelevant: true,
      category,
      keywords: presentKeywords.slice(0, 10)  // Max 10 mots-clés
    };
  }
  
  return { isRelevant: false };
}

/**
 * Sauvegarde les articles dans un fichier TypeScript
 */
function saveToTypeScript(articles) {
  console.log('\n💾 Génération du fichier TypeScript...');
  
  const ssRelevantCount = articles.filter(a => a.relevantForSS).length;
  
  let tsContent = `/**
 * Code Civil Algérien - Tous les articles extraits du PDF
 * Extraction automatique - ${new Date().toLocaleDateString('fr-FR')}
 * 
 * Total: ${articles.length} articles
 * Pertinents pour SS: ${ssRelevantCount} articles
 */

export interface CivilCodeArticle {
  number: string;
  content: string;
  category?: string;
  keywords?: string[];
  relevantForSS: boolean;
}

export const allCivilCodeArticles: CivilCodeArticle[] = [
`;

  articles.forEach(article => {
    // Échapper les caractères spéciaux
    const content = article.content
      .replace(/\\/g, '\\\\')
      .replace(/`/g, '\\`')
      .replace(/\$/g, '\\$');
    
    tsContent += `  {
    number: "${article.number}",
    content: \`${content}\`,
`;
    
    if (article.category) {
      tsContent += `    category: "${article.category}",\n`;
    }
    
    if (article.keywords && article.keywords.length > 0) {
      tsContent += `    keywords: ${JSON.stringify(article.keywords)},\n`;
    }
    
    tsContent += `    relevantForSS: ${article.relevantForSS}
  },
`;
  });

  tsContent += `];

// Articles pertinents pour la sécurité sociale uniquement
export const ssRelevantArticles = allCivilCodeArticles.filter(a => a.relevantForSS);

// Recherche par numéro d'article
export function getArticle(number: string): CivilCodeArticle | undefined {
  return allCivilCodeArticles.find(a => a.number === number);
}

// Recherche par texte
export function searchArticles(query: string): CivilCodeArticle[] {
  const lowerQuery = query.toLowerCase();
  return allCivilCodeArticles.filter(a => 
    a.number.includes(query) ||
    a.content.toLowerCase().includes(lowerQuery) ||
    (a.keywords && a.keywords.some(k => k.toLowerCase().includes(lowerQuery)))
  );
}

// Recherche par catégorie
export function getArticlesByCategory(category: string): CivilCodeArticle[] {
  return allCivilCodeArticles.filter(a => a.category === category);
}

// Statistiques
export const stats = {
  total: ${articles.length},
  ssRelevant: ${ssRelevantCount},
  categories: allCivilCodeArticles.reduce((acc, a) => {
    if (a.category) {
      acc[a.category] = (acc[a.category] || 0) + 1;
    }
    return acc;
  }, {} as Record<string, number>)
};
`;

  fs.writeFileSync(OUTPUT_PATH, tsContent, 'utf-8');
  console.log(`✅ Fichier créé: ${OUTPUT_PATH}`);
}

/**
 * Fonction principale
 */
async function main() {
  console.log('='.repeat(60));
  console.log('🚀 EXTRACTION DU CODE CIVIL ALGÉRIEN DEPUIS PDF');
  console.log('='.repeat(60));
  
  try {
    // Étape 1: Extraire le texte du PDF
    const text = await extractTextFromPDF();
    
    // Étape 2: Extraire les articles
    const articles = extractArticles(text);
    
    if (articles.length === 0) {
      console.log('❌ Aucun article trouvé dans le PDF');
      return;
    }
    
    // Étape 3: Sauvegarder
    saveToTypeScript(articles);
    
    // Résumé
    const ssRelevantCount = articles.filter(a => a.relevantForSS).length;
    const categories = articles
      .filter(a => a.category)
      .reduce((acc, a) => {
        acc[a.category] = (acc[a.category] || 0) + 1;
        return acc;
      }, {});
    
    console.log('\n' + '='.repeat(60));
    console.log('📊 RÉSUMÉ DE L\'EXTRACTION');
    console.log('='.repeat(60));
    console.log(`Total articles: ${articles.length}`);
    console.log(`Articles pertinents SS: ${ssRelevantCount}`);
    console.log('\nRépartition par catégorie:');
    Object.entries(categories).forEach(([cat, count]) => {
      console.log(`  - ${cat}: ${count} articles`);
    });
    console.log('='.repeat(60));
    
  } catch (error) {
    console.error('❌ Erreur:', error);
  }
}

// Exécution
main();

File: scripts/extractAllCivilCodeFromPDF.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script d'extraction de tous les articles du Code Civil Algérien depuis le PDF
Extrait le texte exact tel qu'il apparaît dans le PDF
"""

import PyPDF2
import re
import json
import os

def extract_text_from_pdf(pdf_path):
    """Extrait tout le texte du PDF"""
    print(f"📖 Lecture du PDF: {pdf_path}")
    text = ""
    
    try:
        with open(pdf_path, 'rb') as file:
            pdf_reader = PyPDF2.PdfReader(file)
            total_pages = len(pdf_reader.pages)
            print(f"📄 Nombre de pages: {total_pages}")
            
            for page_num in range(total_pages):
                page = pdf_reader.pages[page_num]
                text += page.extract_text() + "\n"
                
                if (page_num + 1) % 10 == 0:
                    print(f"   Traité: {page_num + 1}/{total_pages} pages")
        
        print(f"✅ Extraction terminée: {len(text)} caractères")
        return text
    
    except Exception as e:
        print(f"❌ Erreur lors de l'extraction: {str(e)}")
        return None

def extract_articles(text):
    """Extrait tous les articles du texte"""
    print("\n🔍 Extraction des articles...")
    
    articles = []
    
    # Pattern pour détecter les articles
    # Format: "Article XXX" ou "Art. XXX" suivi du texte jusqu'au prochain article
    article_pattern = r'(?:Article|Art\.?)\s+(\d+(?:\s*bis|\s*ter)?)\s*[:\.\-]?\s*(.*?)(?=(?:Article|Art\.?)\s+\d+|$)'
    
    matches = re.finditer(article_pattern, text, re.DOTALL | re.IGNORECASE)
    
    for match in matches:
        article_num = match.group(1).strip()
        article_content = match.group(2).strip()
        
        # Nettoyer le contenu
        article_content = re.sub(r'\s+', ' ', article_content)  # Espaces multiples
        article_content = article_content[:2000]  # Limiter à 2000 caractères
        
        if len(article_content) > 10:  # Ignorer les articles vides
            articles.append({
                'number': article_num,
                'content': article_content
            })
    
    print(f"✅ {len(articles)} articles extraits")
    return articles

def categorize_article(article_num, content):
    """Détermine la catégorie et les mots-clés d'un article"""
    content_lower = content.lower()
    
    # Mots-clés pour la sécurité sociale
    ss_keywords = [
        'responsabilité', 'dommage', 'préjudice', 'indemnisation', 'réparation',
        'incapacité', 'invalidité', 'accident', 'travail', 'maladie',
        'professionnel', 'employeur', 'salarié', 'victime', 'consolidation',
        'expertise', 'médical', 'frais', 'soins', 'traitement', 'tierce personne',
        'esthétique', 'moral', 'corporel', 'faute', 'négligence', 'imprudence'
    ]
    
    # Vérifier si l'article est pertinent pour la sécurité sociale
    relevance_score = sum(1 for kw in ss_keywords if kw in content_lower)
    
    if relevance_score >= 2:
        # Catégoriser
        if any(word in content_lower for word in ['responsabilité', 'faute', 'négligence', 'imprudence']):
            category = 'Responsabilité'
        elif any(word in content_lower for word in ['indemnisation', 'réparation', 'dommage', 'préjudice']):
            category = 'Indemnisation'
        elif any(word in content_lower for word in ['expertise', 'consolidation', 'procédure']):
            category = 'Procédure'
        else:
            category = 'Général'
        
        # Extraire les mots-clés présents
        keywords = [kw for kw in ss_keywords if kw in content_lower]
        
        return True, category, keywords
    
    return False, None, []

def save_to_typescript(articles, output_path):
    """Sauvegarde les articles dans un fichier TypeScript"""
    print(f"\n💾 Sauvegarde dans: {output_path}")
    
    ts_content = """/**
 * Articles du Code Civil Algérien
 * Extraction automatique depuis Code_Civil.pdf
 * Date: """ + "19 octobre 2025" + """
 */

export interface CivilCodeArticle {
  number: string;
  content: string;
  category?: string;
  keywords?: string[];
  relevantForSS: boolean;
}

export const allCivilCodeArticles: CivilCodeArticle[] = [
"""
    
    for article in articles:
        # Échapper les guillemets et backslashes
        content = article['content'].replace('\\', '\\\\').replace('"', '\\"').replace('\n', ' ')
        
        ts_content += f"""  {{
    number: "{article['number']}",
    content: "{content}",
"""
        
        if 'category' in article:
            ts_content += f"""    category: "{article['category']}",
"""
        
        if 'keywords' in article and article['keywords']:
            keywords_str = '", "'.join(article['keywords'])
            ts_content += f"""    keywords: ["{keywords_str}"],
"""
        
        ts_content += f"""    relevantForSS: {str(article.get('relevantForSS', False)).lower()}
  }},
"""
    
    ts_content += """];

// Articles pertinents pour la sécurité sociale
export const ssRelevantArticles = allCivilCodeArticles.filter(a => a.relevantForSS);

// Recherche par numéro
export function getArticle(number: string): CivilCodeArticle | undefined {
  return allCivilCodeArticles.find(a => a.number === number);
}

// Recherche par texte
export function searchArticles(query: string): CivilCodeArticle[] {
  const lowerQuery = query.toLowerCase();
  return allCivilCodeArticles.filter(a => 
    a.content.toLowerCase().includes(lowerQuery) ||
    (a.keywords && a.keywords.some(k => k.toLowerCase().includes(lowerQuery)))
  );
}

// Statistiques
export const stats = {
  total: allCivilCodeArticles.length,
  ssRelevant: ssRelevantArticles.length,
  categories: {} as Record<string, number>
};

allCivilCodeArticles.forEach(a => {
  if (a.category) {
    stats.categories[a.category] = (stats.categories[a.category] || 0) + 1;
  }
});
"""
    
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(ts_content)
    
    print(f"✅ Fichier TypeScript créé: {len(articles)} articles")

def main():
    print("=" * 60)
    print("🚀 EXTRACTION DU CODE CIVIL ALGÉRIEN")
    print("=" * 60)
    
    # Chemins
    pdf_path = os.path.join('data', 'Code_Civil.pdf')
    output_path = os.path.join('data', 'civilCodeComplete.ts')
    
    # Étape 1: Extraire le texte du PDF
    text = extract_text_from_pdf(pdf_path)
    if not text:
        print("❌ Impossible d'extraire le texte du PDF")
        return
    
    # Étape 2: Extraire les articles
    articles = extract_articles(text)
    if not articles:
        print("❌ Aucun article trouvé")
        return
    
    # Étape 3: Catégoriser les articles
    print("\n📊 Catégorisation des articles...")
    categorized_articles = []
    ss_relevant_count = 0
    
    for article in articles:
        is_relevant, category, keywords = categorize_article(
            article['number'], 
            article['content']
        )
        
        categorized_articles.append({
            'number': article['number'],
            'content': article['content'],
            'category': category,
            'keywords': keywords,
            'relevantForSS': is_relevant
        })
        
        if is_relevant:
            ss_relevant_count += 1
    
    print(f"✅ Articles pertinents pour SS: {ss_relevant_count}/{len(articles)}")
    
    # Étape 4: Sauvegarder
    save_to_typescript(categorized_articles, output_path)
    
    # Résumé
    print("\n" + "=" * 60)
    print("📊 RÉSUMÉ")
    print("=" * 60)
    print(f"Total articles extraits: {len(articles)}")
    print(f"Articles pertinents SS: {ss_relevant_count}")
    print(f"Fichier généré: {output_path}")
    print("=" * 60)

if __name__ == "__main__":
    main()
